### ISSUE TITLE
Decorator argument types are not checked

### ISSUE THREAD

**Issue by mikeconley12** [microsoft/TypeScript#41029]
**TypeScript Version:**  4.1.0-dev.20201010

**Search Terms:** decorators, ignore `never`, conditional types, mapped types

**Code**

```ts
interface Type<T> extends Function {
  new(...args: any[]): T;
}

type MethodKeys<T> = ({
  [K in keyof T]: T[K] extends Function ? K : never
})[keyof T];

type HookMap<T> = {
  [func in MethodKeys<T>]?: any[];
};

function Hooks<T>(hookMap: HookMap<T>) {
  return (clazz: Type<T>): void => {
    // ...
  };
}

class Parent {
  parentMethod() {
    // ...
  }
}

@Hooks("invalid-value")
class Child extends Parent {
  childMethod() {
    // ...
  }
}
```

**Expected behavior:**
The above code should throw an error because the `"invalid-value"` string isn't assignable to type `HookMap<T>`.

**Actual behavior:**
The `@Hooks()` decorator accepts **any** values without errors.

**Playground Link:** [playground](https://www.typescriptlang.org/play?ts=4.1.0-dev.20201010#code/JYOwLgpgTgZghgYwgAgCoE8AOEA8qB8yEAHpCACYDOyAYgK4gJjAD2IyA3gLABQyyICAHcAFADoJcKAHNKALmRwQ6ANoBdAJQLUAbl4BfXrzBYUAWQhgAFi3IBpCOkp5CAXmQjufZCrvJQyADWjiwwaGravmpEpBAU1PSMzGzIAPzIfgqCAG7QBhoqweih4Xo8xqbIABIsLIFmcJguyO5e-CowDAj+7BbWtg5OLmqpCkqqamX6ZbydSazsNXXOBCI2dQ2YCkv1jS4anLz8UJZ0UOwiCAA2cABet9qm+wrZLMDkLYRt-MgA9L-ICRiI7IaYGIw8a5wSjUAAKUjiYEO3kwCPAfRs5BEB2+-H+gIkIMMPGJvAAAjtKCIAESgbJwK7vAC09KudAg1I0vChMOQAGErMArh8SGQqMh4SdwMj+AhBcKMbZsTKfvigUSDEA)

**Related Issues:**

**Comment by MartinJohns** [comment]
Your example code can be simplified to:

    type MethodKeys<T> = ({ [K in keyof T]: T[K] extends Function ? K : never })[keyof T];
    type HookMap<T> = { [func in MethodKeys<T>]?: any[]; };

    const hooks: HookMap<unknown> = 'invalid-value';

You didn't apply any argument for the type parameter `T`, so the type of `hookMap` is inferred to be `HookMap<unknown>`. And `"invalid-value"` **is** assignable to `HookMap<unknown>`.

You probably expected `T` to be the type the hook is applied to, but this is not supported.

**Comment by mikeconley12** [comment]
Thanks for the explanation!
